### PR TITLE
Add float_datatype parameter for SingleCells compartment load and merge performance flexibility

### PR DESCRIPTION
**Pull request by d33bs**
# Description

This change adds a float_datatype parameter for use with SingleCells compartment load and merge operations to assist with performance considerations (as cited within wayscience/cell-health-data#10 and other areas). In testing this locally using `np.float32` instead of `np.float64` I found ~45% reduction in peak memory consumption while running `merge_single_cells` on small testing datasets (this reduction may vary contingent on dataset).

This change is also related to #198 and #227 (if only to notate datatype considerations).

## What is the nature of your change?

- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: pycytominer/cyto_utils/cells.py [hunk 1]
```diff
@@ -429,13 +429,17 @@ def get_sql_table_col_names(self, table):
...
+    def load_compartment(self, compartment, float_datatype: type = np.float64):
         """Creates the compartment dataframe.
 
         Parameters
         ----------
         compartment : str
             The compartment to process.
+        float_datatype: type, default np.float64
+            Numpy floating point datatype to use for load_compartment and resulting
+            dataframes. Please note: using any besides np.float64 are experimentally
```
Comment (gwaybio) [inline pycytominer/cyto_utils/cells.py:441]: It might be worth noting other options one might use in this docstring and including what the anticipated benefit might be. (i.e. - why is this worth adding as an argument and why anyone would ever change it?)

File: pycytominer/cyto_utils/cells.py [hunk 2]
```diff
@@ -429,13 +429,17 @@ def get_sql_table_col_names(self, table):
 
         return meta_cols, feat_cols
 
-    def load_compartment(self, compartment):
+    def load_compartment(self, compartment, float_datatype: type = np.float64):
```
Comment (gwaybio) [inline pycytominer/cyto_utils/cells.py:432]: I think most folks will use this after initializing the SingleCells class. So it's probably worth making this an init argument and using `self.single_cell_merge_float_datatype` throughout (note the rename - maybe there is something better to rename it to?)

File: pycytominer/tests/test_cyto_utils/test_cells.py [hunk 3]
```diff
@@ -280,6 +281,23 @@ def test_load_compartment():
         check_dtype=False,
     )
 
+    # test using non-default float_datatype
+    loaded_compartment_df = AP.load_compartment(
+        compartment="cells", float_datatype=np.float32
+    )
+    pd.testing.assert_frame_equal(
+        loaded_compartment_df,
+        CELLS_DF.astype(
```
Comment (gwaybio) [inline pycytominer/tests/test_cyto_utils/test_cells.py:290]: This chaining is a bit hard to follow. Can you create a new df (`CELLS_float32_DF` or something), and make sure it is a deep copy? (see https://github.com/cytomining/pycytominer/pull/241#discussion_r1028571680)

File: pycytominer/tests/test_cyto_utils/test_cells.py [hunk 4]
```diff
@@ -280,6 +281,23 @@ def test_load_compartment():
...
+            # cast any float type columns to float32 for expected comparison
+            {
+                colname: np.float32
+                for colname in CELLS_DF.columns
+                if pd.api.types.is_float(CELLS_DF[colname].dtype)
+            }
+        ).reindex(columns=loaded_compartment_df.columns),
+        check_dtype=False,
```
Comment (gwaybio) [inline pycytominer/tests/test_cyto_utils/test_cells.py:298]: Shouldn't `check_dtype=True`? Or maybe i'm not following the chaining.

File: pycytominer/tests/test_cyto_utils/test_cells.py (hunk 3 above)
Comment (gwaybio) [reply pycytominer/tests/test_cyto_utils/test_cells.py:290]: I have the same comment for the second test

File: pycytominer/tests/test_cyto_utils/test_cells.py (hunk 4 above)
Comment (gwaybio) [reply pycytominer/tests/test_cyto_utils/test_cells.py:298]: I have the same comment for the second test

File: pycytominer/tests/test_cyto_utils/test_cells.py [hunk 5]
```diff
@@ -416,6 +434,48 @@ def test_merge_single_cells():
         traditional_norm_df.loc[:, new_compartment_cols].abs().describe(),
     )
 
+    # use non-default float_datatype
+    sc_merged_df = AP.merge_single_cells(float_datatype=np.float32)
```
Comment (gwaybio) [inline pycytominer/tests/test_cyto_utils/test_cells.py:438]: I'm a bit confused why this test is necessary? What are we testing here?

**Review by gwaybio — CHANGES_REQUESTED** [review]
Looks good @d33bs ! This should help @axiomcura in WayScience/cell-health-data#10 (I think this was the right PR to tag here, but it will also likely help many others as well!)

File: pycytominer/cyto_utils/cells.py (hunk 1 above)
Comment (d33bs) [reply pycytominer/cyto_utils/cells.py:441]: Great point, thank you - I'll update this to include a mention about the relationship to performance and an example.

File: pycytominer/cyto_utils/cells.py (hunk 2 above)
Comment (d33bs) [reply pycytominer/cyto_utils/cells.py:432]: This makes a lot of sense, thank you for this feedback. Pondering on this makes me wonder if perhaps more than floating point data (for example, int's) may need or benefit from a default. I'll add in an attribute for `SingleCells` float datatypes with this in mind.

File: pycytominer/tests/test_cyto_utils/test_cells.py (hunk 3 above)
Comment (d33bs) [reply pycytominer/tests/test_cyto_utils/test_cells.py:290]: Understood and will do, thank you for the comment to address this. I'll add an update to this test to make certain it's addressed.

File: pycytominer/tests/test_cyto_utils/test_cells.py (hunk 5 above)
Comment (d33bs) [reply pycytominer/tests/test_cyto_utils/test_cells.py:438]: Thank you for the check here - I added this in to make certain the `float_datatype` work could cascade correctly from `merge_single_cells` in addition to `load_compartment`. With the changes you suggested for the class attribute I think this may no longer be relevant, so I'll remove it.

File: pycytominer/cyto_utils/cells.py (hunk 2 above)
Comment (gwaybio) [reply pycytominer/cyto_utils/cells.py]: > Pondering on this makes me wonder if perhaps more than floating point data (for example, int's) may need or benefit from a default.

Yeah, it likely would! IIRC we had discussed this previously (creating some sort of feature dictionary) but decided to leave this for a future research project. 

My sense is to focus on `float_datatype` here, since we've seen such huge performance boosts and then worry about this dictionary in a future PR

File: pycytominer/tests/test_cyto_utils/test_cells.py (hunk 5 above)
Comment (gwaybio) [reply pycytominer/tests/test_cyto_utils/test_cells.py]: got it!

It's probably sufficient to assert `self.float_datatype` once you make the class attribute change

File: pycytominer/tests/test_cyto_utils/test_cells.py [hunk 6]
```diff
@@ -1050,3 +1079,12 @@ def test_load_non_canonical_image_table():
         result.sort_index(axis="columns").drop("Metadata_Site_Count", axis="columns"),
         sc_aggregated_df,
     )
+
+def test_singlecells_default_datatype():
```
Comment (gwaybio) [inline pycytominer/tests/test_cyto_utils/test_cells.py:1083]: this can probably be deleted, right?

File: pycytominer/tests/test_cyto_utils/test_cells.py [hunk 7]
```diff
@@ -293,6 +294,42 @@ def test_load_compartment():
...
+    ).load_compartment(compartment="cells")
+
+    # for uniformly handling metadata types for both dataframes
+    metadata_types = {"ObjectNumber": "int64"}
+
+    # create deep copy of CELLS_DF with manually re-typed float columns as float32
+    cells_df_for_compare = CELLS_DF.copy(deep=True).astype(
+        # cast any float type columns to float32 for expected comparison
+        {
+            colname: np.float32
```
Comment (gwaybio) [inline pycytominer/tests/test_cyto_utils/test_cells.py:310]: is this blacked? I've never seen this format before, but I can read it!

**Review by gwaybio — APPROVED** [review]
Looks good @d33bs ! Only two minor comments but then I think we're good to merge.

File: pycytominer/tests/test_cyto_utils/test_cells.py (hunk 6 above)
Comment (d33bs) [reply pycytominer/tests/test_cyto_utils/test_cells.py:1083]: Thank you, yes, apologies. I'll remove this block.

File: pycytominer/tests/test_cyto_utils/test_cells.py (hunk 7 above)
Comment (d33bs) [reply pycytominer/tests/test_cyto_utils/test_cells.py:310]: I did apply black styling to this, but admittedly the dictionary comprehension mid-block makes things look maybe a bit unusual. Alternatively we could move the dictionary before the `astype` to make this a bit less terse or better matching the reading style of the testing suite. Thoughts?

Comment (gwaybio) [reply pycytominer/tests/test_cyto_utils/test_cells.py:310]: My preference is to split out the dictionary compensation to increase readability. Love the comments though, so please retain these!

Comment (d33bs) [reply pycytominer/tests/test_cyto_utils/test_cells.py:310]: Thanks, sounds good, will do.

**Comment by gwaybio** [comment]
Looks good! I'm going to merge. Thanks @d33bs
